### PR TITLE
CosmosClientOptions: Adds validation to RequestTimeout to block values under 1 second

### LINKED ISSUE #958: Unclear error when RequestTimeout set to below 1 second
**Describe the bug**
If setting the `RequestTimeout` on `CosmosClientOptions` to less than 1 second, attempts to call cosmos fail with unclear errors.

**To Reproduce**
If you try the below code with a `999`ms timeout, it fails with the exception below.  Bump it to `1000` and it starts working as expected.
```cs
var options = new CosmosClientOptions
{
    RequestTimeout = TimeSpan.FromMilliseconds(999)
};

var client = new CosmosClient(DatabaseUri, AuthKey, options);
await client.ReadAccountAsync();
```

**Expected behaviour**
Given that Cosmos provides it's SLA for certain operations at 10ms or 15ms, I'd expect to be able to set timeouts much more granular than 1 second.

If timeouts less than 1 sec are not allowed, i'd expect a clearer error stating that the timeout must be at least 1 second, rather than the ambiguous exception currently given:

**Actual behavior**
```
Process terminated. Assertion failed.
   at Microsoft.Azure.Documents.Rntbd.TransportClient.Options..ctor(TimeSpan requestTimeout)
   at Microsoft.Azure.Documents.StoreClientFactory..ctor(Protocol protocol, Int32 requestTimeoutInSeconds, Int32 maxConcurrentConnectionOpenRequests, UserAgentContainer userAgent, ICommunicationEventSource eventSource, String overrideHostNameInCertificate, Int32 openTimeoutInSeconds, Int32 idleTimeoutInSeconds, Int32 timerPoolGranularityInSeconds, Int32 maxRntbdChannels, Int32 rntbdPartitionCount, Int32 maxRequestsPerRntbdChannel, TcpPortReuse rntbdPortReusePolicy, Int32 rntbdPortPoolReuseThreshold, Int32 rntbdPortPoolBindAttempts, Int32 receiveHangDetectionTimeSeconds, Int32 sendHangDetectionTimeSeconds, Boolean enableCpuMonitor, RetryWithConfiguration retryWithConfiguration)
   at Microsoft.Azure.Cosmos.DocumentClient.InitializeDirectConnectivity(IStoreClientFactory storeClientFactory)
   at Microsoft.Azure.Cosmos.DocumentClient.GetInitializationTaskAsync(IStoreClientFactory storeClientFactory)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at Microsoft.Azure.Cosmos.DocumentClient.InitializeGatewayConfigurationReaderAsync()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at Microsoft.Azure.Cosmos.CosmosAccountServiceConfiguration.InitializeAsync()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at Microsoft.Azure.Cosmos.GatewayAccountReader.InitializeReaderAsync()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at Microsoft.Azure.Cosmos.Routing.GlobalEndpointManager.GetDatabaseAccountFromAnyLocationsAsync(Uri defaultEndpoint, IList`1 locations, Func`2 getDatabaseAccountFn)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at Microsoft.Azure.Cosmos.GatewayAccountReader.GetDatabaseAccountAsync(Uri serviceEndpoint)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at Microsoft.Azure.Cosmos.DocumentClient.HttpRequestMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetExistingTaskResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(IAsyncStateMachineBox box, Boolean allowInlining)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder`1.SetResult(TResult result)
   at System.Net.Security.SslStream.<FillBufferAsync>g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Threading.ThreadPoolGlobals.<>c.<.cctor>b__5_0(Object state)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.InvokeContinuation(Action`1 continuation, Object state, Boolean forceAsync, Boolean requiresExecutionContextFlow)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.OnCompleted(SocketAsyncEventArgs _)
   at System.Net.Sockets.SocketAsyncEventArgs.<>c.<.cctor>b__177_0(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading.ThreadPoolBoundHandleOverlapped.CompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* nativeOverlapped)
   at System.Threading._IOCompletionCallback.PerformIOCompletionCallback(UInt32 errorCode, UInt32 numBytes, NativeOverlapped* pNativeOverlapped)

C:\...\repos\ConsoleApp3\bin\Debug\netcoreapp3.0\ConsoleApp3.exe (process 17076) exited with code -2146232797.
Press any key to close this window . . .
```

**Environment summary**
SDK Version: 3.3.3
OS Version: Microsoft Windows [Version 10.0.18362.387]

### PR DESCRIPTION
**Pull request by SchintaMicrosoft**
# Pull Request Template

## Description
Validation added for Request Timeout to check Request Timeout cannot be set less than 1 second. This must be set in the Timespan of seconds.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #958

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 1]
```diff
@@ -156,9 +157,22 @@ public int GatewayModeMaxConnectionLimit
...
         /// </summary>
-        /// <value>Default value is 1 minute.</value>
+        /// <value>Default value is 10 seconds.</value>
         /// <seealso cref="CosmosClientBuilder.WithRequestTimeout(TimeSpan)"/>
-        public TimeSpan RequestTimeout { get; set; }
+        public TimeSpan RequestTimeout 
+        {
+            get => this.requestTimeout;
+
+            set
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:166]: Every change should add tests to cover what is being changed. Please add tests.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 2]
```diff
@@ -156,9 +157,22 @@ public int GatewayModeMaxConnectionLimit
...
+            {
+                if (value != null && value < TimeSpan.FromSeconds(1))
+                {
+                    throw new ArgumentException($"{nameof(this.requestTimeout)} can not be set less than 1 second. This must be set in the Timespan of seconds.");
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: nit: We should use the public property name
```suggestion
                    throw new ArgumentException($"{nameof(this.RequestTimeout)} can not be set less than 1 second. This must be set in the Timespan of seconds.");
```

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 3]
```diff
@@ -156,9 +157,22 @@ public int GatewayModeMaxConnectionLimit
...
+                }
+
+                this.requestTimeout = value;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:173]: nit: Should we validate that the TimeSpan has no milliseconds? and that is positive? Like in https://github.com/Azure/azure-cosmos-dotnet-v3/pull/2991/files

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 4]
```diff
@@ -156,9 +157,22 @@ public int GatewayModeMaxConnectionLimit
...
-        public TimeSpan RequestTimeout { get; set; }
+        public TimeSpan RequestTimeout
+        {
+            get => this.requestTimeout;
+
+            set
+            {
+                if (value.Seconds < 0)
+                {
+                    throw new ArgumentOutOfRangeException($"{nameof(this.RequestTimeout)} can not be set less than 1 second. This must be set in the Timespan of seconds.");
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: Check and error message are out of sync - code only would throw with negative timespan. error message indicates it would throw even when seconds == 0

Also this is a breaking change - throwing Argument Exceptions where it didn't before. If it is scoped to just negative timeouts it owuld be good to explain what is happening before this PR and explain why it is ok to break. If the plan is to throw any exception for postive Tiemspan i think this breaking change is unacceptable.

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Breaking change

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: The SDK did throw an exception from version 3.0 to 3.6. The check got removed because of some changes in the direct layer. This is not a breaking change. It's simply adding back the original check.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 3 above)
Comment (bartelink) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:173]: Please don't reject non-zero milliseconds - I would not be surprised if people use values like 4.5 or 5.5s, and this logic can often be buried in long-established deep config paths which won't do useful error logging and hence be hard to troubleshoot. (if the granularity was truly seconds, then also consider a `int timeoutS` representation)

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: It is still a breaking change if there is any minor version in the same 3.* major version that did not throw. Removing exceptions from a public API is a one-way-street, you can't revert that decision without moving to a new major version.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: How is adding validation for an invalid value a breaking change? Throwing an invalid argument exception on a config value should be expected. It will be thrown on the set so it is a fast failure with a clear error. If the customer does any testing or validation, it will fail and they can actually fix their application. This seems to be a far better user experience then ignoring the value which will eventually lead to a live site issue because they built their app on the assumption, they could do a request timeout of 100ms.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 3 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:173]: Normally I would say that I rather have an exception than false expectations. But in this case, this is an API that already existed, and on the configuration path (possibly a place where the client is created as part of the application initialization). Any new exceptions being thrown there would cause the application to restart/enter a cycle that the user might not know how to recover from ("I just updated my SDK and now my app does not start, I was simply setting timeout as 1.356 seconds").

If this was a new API introduced, then yeah, the check makes sense.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (SchintaMicrosoft) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: @FabianMeiswinkel , @j82w any further suggestions here  or proceed with this change.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 5]
```diff
@@ -156,9 +157,22 @@ public int GatewayModeMaxConnectionLimit
...
+        /// <value>Default value is 10 seconds.</value>
         /// <seealso cref="CosmosClientBuilder.WithRequestTimeout(TimeSpan)"/>
-        public TimeSpan RequestTimeout { get; set; }
+        public TimeSpan RequestTimeout
+        {
+            get => this.requestTimeout;
+
+            set
+            {
+                if (value.Seconds < 0)
```
Comment (bartelink) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:168]: should be TotalSeconds

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (bartelink) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]: use overload with argument name, don't embed it in the message. Lose the second half of the message as you have removed that semantic

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 5 above)
Comment (bartelink) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:168]: You probably mean `<1` per the PR title. 

The overview does not provide a clear message as to the goal; I feel that getting that right will make it easier to determine whether an exception message does or does not concisely convey the intent.

I believe this has been traversed elsewhere in this PR review but arbitrarily saying you can't make the timeout be 5.1 seconds is definitely a bad idea (I personally don't see whats magic about 1s vs 500ms either in the context of a call to Cosmos in the same data center given a typical request latency of <100ms for the bulk of requests)

Long story short, please don't get distracted by picky comments on the impl - IMO the high level intent is not clear and that should be established first.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 6]
```diff
@@ -407,6 +407,16 @@ public void VerifyCorrectProtocolIsSet()
...
             Assert.AreEqual(Protocol.Tcp, cosmosClientOptions.ConnectionProtocol);
         }
+        [TestMethod]
+        public void VerifyRequestTimeoutToEndpointSettings()
+        {
+            CosmosClientOptions cosmosClientOptions = new CosmosClientOptions { ConnectionMode = ConnectionMode.Gateway };
+            cosmosClientOptions = new CosmosClientOptions { RequestTimeout = TimeSpan.FromSeconds(1) };
+            Assert.AreEqual(TimeSpan.FromSeconds(1), cosmosClientOptions.RequestTimeout);
+
+            cosmosClientOptions = new CosmosClientOptions { RequestTimeout = TimeSpan.FromMilliseconds(500) };
```
Comment (bartelink) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:417]: so 500ms is allowed? Or not?

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:170]:
```suggestion
                    throw new ArgumentOutOfRangeException(nameof(this.RequestTimeout), "Timeout can not be set less than 1 second. This must be set in the Timespan of seconds.");
```

**Review by ealsur — APPROVED** [review]
Please sync with Fabian and get clarification if we want to move forward with the new exception thrown. cc @kirankumarkolli

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 7]
```diff
@@ -156,9 +157,22 @@ public int GatewayModeMaxConnectionLimit
         /// Gets the request timeout in seconds when connecting to the Azure Cosmos DB service.
         /// The number specifies the time to wait for response to come back from network peer.
         /// </summary>
-        /// <value>Default value is 1 minute.</value>
+        /// <value>Default value is 10 seconds.</value>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs:160]: [Unrelated to current PR]
@ealsur do default 10 seconds apply for gateway meta-data calls? If so will it impact fail-overs?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs (hunk 6 above)
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs:417]: With 500ms won't this line throw exception?

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs (hunk 4 above)
Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Semantically what @FabianMeiswinkel said is true. Post 3.6 the customers might have used values which are invalid but silently ignored now so it might break such customers. 

Its better to be very explicit on what is not supported. 
How about documenting it clearly in release notes as possible impact?

Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: I am against this breaking change - what is the motivation for arguing that this breaking change is worth violating semantic versioning? How many CRIs have we seen because customers were confused that their timeouts < 1 second haven't been accepted. And if they are confused they won't like either that they can't set those shorter timeouts. So, either you open the flood gate and allow customers to set whatever timeout you want as real TimeSpan  (taking the risk that they shoot themselves in the foot and end-up with bad availability) or you keep the current behavior and document it as clearly as possible in API doc - but this breaking change is not helping much and is still causing the pain breaking changes can have.

Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: Closing this PR unmerged is better than taking this change IMO. At least in V3

Comment (bartelink) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: I would tend to round up or down rather than throw an exception as config wiring like this will often not traverse exception paths, much less be covered in tests IRL.
>`This must be set in the Timespan of seconds.`
If it is to proceed, perhaps a better wording is `The Timespan must be in whole Seconds.`

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: There is consensus on not introducing the breaking change. 
Let's abandon this PR.

**Comment by SchintaMicrosoft** [comment]
Abandoning this PR because of breaking charge.
